addons/stock/models/stock_inventory.py

```diff
@@ -460,10 +430,45 @@ def _generate_moves(self):
         for line in self:
             if float_utils.float_compare(line.theoretical_qty, line.product_qty, precision_rounding=line.product_id.uom_id.rounding) == 0:
                 continue
-            diff = line.theoretical_qty - line.product_qty
-            if diff < 0:  # found more than expected
-                vals = line._get_move_values(abs(diff), line.product_id.property_stock_inventory.id, line.location_id.id, False)
+            if line.difference_qty > 0:  # found more than expected
+                vals = line._get_move_values(line.difference_qty, line.product_id.property_stock_inventory.id, line.location_id.id, False)
             else:
-                vals = line._get_move_values(abs(diff), line.location_id.id, line.product_id.property_stock_inventory.id, True)
+                vals = line._get_move_values(abs(line.difference_qty), line.location_id.id, line.product_id.property_stock_inventory.id, True)
             vals_list.append(vals)
         return self.env['stock.move'].create(vals_list)
+
+    def _refresh_inventory(self):
+        if self and self[0].inventory_id:
+            return self[0].inventory_id.action_open_inventory_lines()
+
+    def action_refresh_quantity(self):
+        for line in self:
+            if line.outdated:
+                quants = self.env['stock.quant']._gather(line.product_id, line.location_id, lot_id=line.prod_lot_id, package_id=line.package_id, strict=True)
+                if quants.exists():
+                    quantity = sum(quants.mapped(lambda q: q.quantity))
+                    if line.theoretical_qty != quantity:
+                        line.theoretical_qty = quantity
+                line.inventory_date = fields.Datetime.now()
+                line.outdated = False
+        self._refresh_inventory()
+
+    def action_reset_product_qty(self):
+        self.write({'product_qty': 0})
+        self._refresh_inventory()
+
+    @api.model
+    def action_validate_inventory(self, inventory_id):
```

Comment:
You should make this method not 'api.model' and only pass self as parameter. You will avoid a useless browse